### DISCUSSION TITLE
Decoupling Storage and Deal with independent POW and POS mechanisms

### DISCUSSION THREAD

**Discussion by robertl0705** [filecoin-project/FIPs#442] [Enhancements - Technical]
# Summary

This proposal is far from complete but is aiming to spark discussions and ideas. The outstanding suggestion is to decouple the storage and the pledge collaterals with the independent POW and POS-like mechanisms respectively, which can be realized through the following,

- The POW mechanism works for storage providing and block rewards as it currently does. SPs’ (Storage Providers’) block rewards depend on RawBytes (physically proved storage contributed to the network) instead of QAP (Quality Adjusted Power);

- The POS mechanism works for the pledge collaterals, of which rewards are determined by the amount of FIL collateralized. 

- For any SP, its total rewards consist of the POW/Storage rewards and the POS/Collateral rewards, as the SP always needs to pledge no matter commits in CC, normal deals or verified deals.

- There are two categories of Pledge under the POS mechanism, the Sector Pledge and the Deal Pledge. Sector Pledge is calculated with the Sector RawByte power, and Deal Pledge is defined by the subscribed Deal size. 

- Distribute the network total rewards to the POW and the POS components with a fixed or dynamic adjusted proportion (to be discussed).

# Background

In Filecoin network, security is guaranteed by Storage Providers (miners) through continually proving storage capacity, and as an exchange, miners are compensated with block rewards. EC consensus initiates a leader election every epoch in which, an expected number (currently 5) of the participants may be eligible to submit a block and get rewarded. The chance of winning a block is proportional to the Power that miners are processing. 
 
F+ (Filecoin Plus) program was designed to incentivize the storage of useful data with 10 times storage power awarded to the verified deals. The Power used for reward and pledge collateral calculations is not the RawBytes but the weighted average Power of the verified deals in a sector, which is known as the QAP (Quality Adjusted Power). Intuitively, a storage provider that subscribes F+ deals should account for a higher QAP, in other words, higher probability of winning a block. QAP mechanism spreads the deal’s size over the sector’s lifetime, regardless of the deal term. It couples the deal market to the storage sector through duration and rewards. 

# Motivation

Under the current design, lack of flexibility stagnates the development on both sector and deal. Data in the sectors cannot be removed or replaced even if deals are expired. Deals cannot be moved between sectors and duration can hardly be extended. Highly anticipated FVM is expected to facilitate the execution of smart contracts in the Filecoin network, which would promote the innovation of the storage deal market. But without decoupling the concepts of sector and deal, innovative products such as duration extension, inter-sectoral deals, deal multiples, content insurance, collateral backed securities and other financial derivatives can be difficult to implement.
 
In addition, imbalance exists among large and small SPs. SPs with large amount of FIL holdings will actively participate in verified deals, however small SPs who only participate in CC storage or normal deals, which is equally critical to the consensus and security, will be disadvantaged along with its low chance of block winning. The distressed CC miners may leave the network ultimately and compromise the decentralization.  
 
Heavily inspired by [FIP-0033](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0033.md), which proposed to remove the concept of QAP from the storage powered consensus to decouple the concept of the sector quality from the consensus, we can’t help thinking to further simplify the system and break the linkage between the sector and deal by running independent POW and POS mechanisms and redefining pledge and rewards.

# Specification

The POW component guarantees the network security, EC (Expected Consensus) and the rewards of consensus miners as it currently does. For each epoch, a leader election is running to select a set number of SPs (by expectation) to submit a block. EC guarantees that these winners will be anonymous until they reveal themselves by submitting the ElectionProof. Each winning miner will be rewarded proportionally to its RawByte power (instead of the QAP). The consensus is secured by the block rewards yet to be vested and the Sector Initial Pledge which includes, the Sector Initial Storage Pledge and the Sector Initial Consensus Pledge. 
 
The POS component enables rewards for the SPs or any other participants who can put collaterals into the network, regardless of the deal type, CC, normal deals or verified deals. The rewards are determined by the amount of FIL collateralized in relation to total amount of FIL collateralized in the network, a POS-like mechanism. For example, verified deals would pledge 10 times as much as CC or normal deals for the same RawByte sector, therefore earn 10 times as much for the POS rewards, which should have no impact to the POW rewards, as the proportions of POW and POS rewards are pre-determined in a fixed or dynamic adjusted manner. 
 
Conceptually, Sector is the basic unit of storage on Filecoin, which can be seen as the “container”, measured by RawByte Power, whereas a Deal is formed only when a storage miner and client are matched on data storage market, the “infilling”, measured by the deal size. The Sector Pledge and the Deal Pledge should be redefined to organically distinguish the two. One and probably the simplest approach is to calculate the Sector Pledge with the RawBytes and calculate the Deal Pledge with the deal size. A multiple can apply to the deal size if it is a verified deal or other innovative deal products.

**POW Reward can be defined as**
 
$$Total POW Reward = Total Reward \times \alpha$$

<p align="center">
α : fixed or dynamically adjusted proportion of POW reward over the Total Reward
</p>

POW reward for each sector depends on

$$P(Wining a block) \sim Poisson(\lambda)$$

$$\lambda = \frac{Sector RawByte Power}{network RawByte Power} \times 5$$

**POS Reward can be defined as**

$$Total POS Reward = Total Reward \times (1-\alpha)$$

$$POS Reward for Sector= Total POS Reward \times \frac{Sector Pledge}{Total Network Pledge Collateralized}$$

$$POS Reward for Deal= Total POS Reward \times \frac{Deal Pledge}{Total Network Pledge Collateralized}$$

**Sector Pledge can be calculated as**

$$SectorPledge =  0.3 \times FILCirculationSupply \times \frac{SectorRawByte \times 1^*}{max(NetworkBaseline, NetworkRawByte)}$$

<p align="center">
*Pledge Collateral Multiplier for Sector Pledge is always 1
</p>

**Deal Pledge can be calculated as**

$$DealPledge =   0.3  \times FILCirculationSupply \times \frac{Deal Size \times (\beta-1)}{max(NetworkBaseline, NetworkRawByte)}$$

<p align="center">
β : Pledge Collateral Multiplier for different types of deal, as currently F+ = 10
</p>

# Rationale

A snapshot of the proposed structure is as follows,

<p align="center">
<img width="468" alt="image" src="https://user-images.githubusercontent.com/29261438/186124729-b7938f1f-c92b-4dee-b80d-ea3274e58570.png">
</p>

<p align="center">
<img width="468" alt="image" src="https://user-images.githubusercontent.com/29261438/186125693-87ba303b-0fdd-4b0c-ba03-ac9456141eb0.png">
</p>

The advantages of the proposed settings are,

- The independent POW and POS mechanisms ensure that the storage rewards and the collateral rewards are calculated separately. A portion of rewards for storage miners (POW) is secured and is only determined by the RawBytes storage power provided, so that small consensus miners will not be disadvantaged and are able to focus on providing storage and security;

- Dual-mechanism extricates sectors from deals, which as mentioned earlier, would encourage the design of innovative products such as loans, storage insurance, deal delegations, collateral backed securities and other financial derivatives on FVM.

- The concepts of Sector and Deal are decoupled, so that sectors become fungible and deals can be applied to any portion of the sector or moved from on to another. 

- The bonus of this flexibility is that supplemental security is guaranteed by the POS mechanism in addition to the POW consensus. The importance or authenticity of data can be determined by deal participants/subscribers. For example, if deal participants deem necessary, they could subscribe for longer duration or higher collateral multiplier to secure the data stored in the sector and, by nature, earn higher multiple of POS rewards.

# Implementation
This proposal can be seen as a simplification process, which involves change of key concepts and is expected to be discussed, revised and accepted before implementation.

**Comment by anorth** [comment]
@robertl0705 thanks for starting this discussion. I just want to flag that I'm interested in it, but won't be able to engage deeply for the next month or so while focussed on nv17.

**Reply by robertl0705** [reply]
Thanks Alex, good to hear that you are interested in. Let's see how people think about this idea and we can collaborate in a later date.

**Comment by steven004** [comment]
Thanks @robertl0705 for coming up with these ideas. It is inspiring. I am in support of this though lots of details might need to figure out. 

I love the PoW, based on the real-world physical storage contribution, for building the concrete base of consensus, and PoS, simply based on the collateral, for providing a flexible place for innovation. We could think of different kinds of deals with different collateral constrains (multiplier for an example), or let user-define collateral (more than a minimal requirement) for a piece for more important content, or even letting holders directly put collateral for a storage pool as insurance. 

In addition, I would suggest a fixed ratio between PoW and PoS rewards, e.g. 50:50.

**Reply by robertl0705** [reply]
Thanks, Steven! It's always the best to give POS more flexibility and at the same time guarantee the consensus of POW.

**Comment by jimcray** [comment]
Yes, with the storage capacity is increasing, it should be split. The proof of stake should be based on the data market value, not just the size.

**Reply by robertl0705** [reply]
Right, Jim, as the proof of work should be based on the physical storage provided.

**Comment by DecentCr8** [comment]
Good proposal. It is on the right direction, but not thorough and comprehensive thinking.

In a blockchain system, consensus security comes first. Filecoin's design does a very good job of consensus security with some significant innovations. Unfortunately, due to the development of the Fil+ project and the discussion of SDM in the Filecoin system, Filecoin has a tendency to turn to PoS, that is to say, the consensus depends on the mortgage, rather than the real-world network contribution.

According to PoW + PoS defined in the proposal, actually Filecoin only includes the part of PoS right now, that is to say PoW: PoS = 0:100. Fortunately, because the proportion of Verified Deals is still very small, the consensus looks like PoW. However, if the move of Fil+ accelerates, and if SDM is really introduced, then Filecoin will inevitably move towards PoS. This will be a sad end for most Storage Providers.

I would like to see Filecoin Layer 1 go to completely PoW, which mean PoW:PoS=100:0 for the long-term goal of Filecoin, although I admit this is a good proposal and it's on the right direction. That is, for long run, the block rewards only depends on RawByte power to ensure consensus security. Someone asked, how to incentivize real data storage? The answer is, **to Layer2**. 

What we need to do are: 
- Simplify Layer 1 by removing QAP;
- Move Fil+, deals to Layer 2, implemented via contracts; Just define Layer 2 FilPlus Token to incentive instead of using FIL of Layer 1;
- Allow users to establish different Markets and compete among Markets, and the results of the competition are reflected by the Market Capital of the corresponding Actor (Token);
- When the storage incentive for real data is moved to Layer 2, there is absolutely no need for SDM to be discussed at Layer 1. User-defined Markets can implement SDM and be linked to their Layer 2 pledges without needs of discussion (anyone can do it as she wants).

**Reply by robertl0705** [reply]
Thanks for your interest! I 100% agree with you that the network running towards a 0:100 POW:POS could be a sad end, which is why I propose the decoupling with a fixed reward proportion alpha. In that way, the leverage on POS component will not flood the POW mechanism (hopefully). But it is undeniable that the POS mechanism is helpful to secure the network in addition to POW and the minders need to pledge anyways. One key thing I may not make myself clear is that deal and sector should also be decoupled, no matter under Layer 1 or under your excellent solution, Layer 2. I see your proposal an extension of mine with an refined execution plan if you agree.

**Comment by AxCortesCubero** [comment]
@robertl0705 This is an interesting idea, that I think is very similar to something @anorth had brought up before, so I want to take a moment to understand if this is the same thing, and if anorth's original motivations are still valid.

So @anorth I know your original motivation was about separating the FIL+ mechanism into a different actor, to improve FVM related programability. I know this evolved into several different versions of the idea, and its latest version might be FIP 0045? Does FIP 0045 solve all your problems or does your initial motivation still stands and you also think this "FIL+ subsidy" kind of mechanism is still needed?

Back then I had produced [this calculation](https://hackmd.io/@R02mDHrYQ3C4PFmNaxF5bw/SJ61McoOt), basically thinking of the parameter "alpha" described here, how the raw-bite vs FIL+ components of the reward can be split. There is no one single answer, but I would recommend the last option here, which is based on never needing any SP to provide any refund of their reward.

This is not included here, but was part of anorth's original idea that, the FIL+ component, being separated from consensus mechanism, is free to be a smooth reward (you can receive your proportion of reward at every block) and not lottery based. I have reasons to believe this can be beneficial as a cryptoeconomic incentive, to have smooth rather than lottery based reward. So even if @anorth's motiviation is already resolved by FIP 0045, the smooth reward aspect may be enough of a motivation in itself to push this idea.

Lastly a big change in this mechanism is changing the consensus from QAP based to Raw power based. While the explicit block reward is not changed, block reward is still QAP based, this is still an important change in incentives. There is economical value in consensus power as well, for instance miner tips are involved, as well as choice of which messages  and in what order to write in the block (miner extractable value). So this would still be a large change in incentives to think very carefully about.

**Reply by anorth** [reply]
FIP-0033 had a number of motivations. FIP-0045 resolves the most pressing one, but a number of other valid motivations remain and can be met by a proposal like this. I would like to see it redrafted to focus on only the remaining issues.

> Data in the sectors cannot be removed or replaced even if deals are expired. Deals cannot be moved between sectors and duration can hardly be extended. 

This is resolved by FIP-0045. Rather than remove the QAP mechanism entirely, FIP-0045 changes the calculation so as to no longer inhibit updateability of storage. The timing of the reward is also fixed to match the period of service.

Motivations cited for FIP-0033 that are _not_ yet resolved include:
- Uneven and unreliable reward for hosting verified deals, due to probabilistic and operationally-complex nature of block reward
- Undermining of proof of (useful) work arguments for consensus security based around the difficulty of amassing physical hardware
- Coupling of the distinct concerns of network security and reward distribution policy

That proposal included brief discussion of the possibility of adding a parameter to cap the fraction of reward (and, with the current system, consensus power) allocated to the deal premium. I would still like to see such a parameter introduced, especially if we don't decouple these concerns so that, even as verified storage utilisation grows, we can make a long term argument about the physical difficulty (rather than mere financial cost) of maintaining a large fraction of consensus power.
